doc/designs/quic-design/synthesis.md

```diff
@@ -0,0 +1,507 @@
+ * any such call is a no-op.
+ *
+ * For QUIC:
+ *   Creates a new stream. Can be used on client or server. If the SSL_FLAG_UNI
+ *   flag is set, the created stream is unidirectional, otherwise it is
+ *   bidirectional.
+ *
+ * For TLS and DTLS clients:
+ *   Attempts a resumption handshake. The flags argument is ignored.
+ *
+ * For TLS and DTLS servers:
+ *   Always fails.
+ */
+#define SSL_FLAG_UNI    1
+SSL *SSL_new_stream(SSL *ssl, uint64_t flags);
+
+/*
+ * Create a new SSL object representing an additional stream which has created
+ * by the peer.
+ *
+ * There is no need to call SSL_accept on the resulting object, and
+ * any such call is a no-op.
+ *
+ * For QUIC:
+ *   Checks if a new stream has been created by the peer. If it has, creates a
+ *   new SSL object to represent it and returns it. Otherwise, returns NULL.
+ *
+ * For all other methods:
+ *   Returns NULL.
+ *
+ * The flags argument is unused and should be set to zero.
+ *
+ * TODO: Do we want a WAIT flag for synchronous blocking here?
+ */
+SSL *SSL_accept_stream(SSL *ssl, uint64_t flags);
+
+/*
+ * Determine the number of streams waiting to be returned on a subsequent call
+ * to SSL_accept_stream. If this returns a non-zero value, SSL_accept_stream is
+ * guaranteed to work. Returns 0 for non-QUIC objects, or for QUIC stream
+ * objects.
+ *
+ * Not strictly necessary but probably convenient.
+ */
+int SSL_get_accept_stream_queue_len(SSL *ssl);
+
+/*
+ * Set a SSL_STREAM_TYPE value which determines the type of the default stream.
+ * The default value for QUIC methods is SSL_STREAM_TYPE_BIDI.
+ *
+ * This must be set before connecting. It is an error to try to set it to
+ * anything other than SSL_STREAM_TYPE_BIDI for non-QUIC methods (or
+ * we could just have this always fail on non-QUIC methods).
+ *
+ * If this is set to SSL_STREAM_TYPE_NONE, no default stream is created
+ * at connection time and the initial connection object does not represent
+ * a stream. Calls to SSL_read()/SSL_write() on this object fail. Streams
+ * must be created using SSL_new_stream() or SSL_accept_stream().
+ */
+__owur int SSL_set_default_stream_type(SSL *ssl, int type);
+
+/*
+ * Detaches a default stream from a QUIC connection object. If the
+ * QUIC connection object does not contain a default stream, returns NULL.
+ * After calling this, calling SSL_get_stream_type on the connection object
+ * returns SSL_STREAM_TYPE_NONE. Always returns NULL for non-QUIC connections.
+ *
+ * NOTE: This is not really necessary but I imagine some people might find it
+ *       convenient.
+ */
+SSL *SSL_detach_stream(SSL *ssl);
+
+/*
+ * Attaches a default stream to a QUIC connection object. If the conn object is
+ * not a QUIC connection object, or already has a default stream, this function
+ * fails. The stream must belong to the same connection, or this function fails.
+ *
+ * NOTE: This is not really necessary but I imagine some people might find it
+ *       convenient.
```

Comment:
IMO this is unnecessary and should not be added unless we have a clear use case.